### PR TITLE
Gin link installs module

### PR DESCRIPTION
**Pull request by branflake2267**
Progress report... Got the gin module link source getting added to the parent gin module. Although, its not formatting correctly yet, its functionally working. 

screen shot 2013-10-28 at 5 48 33 pm

Up next formatting of the module import statement and method install order.

### CODE REVIEW

File: src/com/arcbees/plugin/idea/utils/PackageHierarchy.java [hunk 1]
```diff
@@ -141,65 +145,67 @@ public boolean isParentTheClientPackage(String packageElementName) {
...
+        return foundUnit;
+    }
+
+    public PsiClass findInterfaceTypeInParentPackage(PsiPackage packageSelected, String findTypeName) {
+        PsiClass[] units = packageSelected.getClasses();
+
+        for (PsiClass unit : units) {
+            boolean found = findInterfaceUseInUnit(unit, findTypeName);
+            if (found == true) {
+                return unit;
```
Comment (spg) [inline src/com/arcbees/plugin/idea/utils/PackageHierarchy.java]: only 1 return statement per method is better

File: src/com/arcbees/plugin/idea/utils/PackageHierarchy.java [hunk 2]
```diff
@@ -141,65 +145,67 @@ public boolean isParentTheClientPackage(String packageElementName) {
...
+    }
+
+    private boolean findInterfaceUseInUnit(PsiClass unit, String findTypeName) {
+        for (PsiClassType types : unit.getSuperTypes()) {
+            PsiType[] superTypes = types.getSuperTypes();
+
+            if (superTypes != null) {
+                for (PsiType superType : superTypes ) {
+                    if (superType.getCanonicalText().contains(findTypeName)) {
+                        return true;
```
Comment (spg) [inline src/com/arcbees/plugin/idea/utils/PackageHierarchy.java:216]: idem

File: src/com/arcbees/plugin/idea/wizards/createpresenter/CreatePresenterAction.java [hunk 3]
```diff
@@ -132,18 +143,146 @@ private void run() {
...
+            name = createdPopupPresenterTemplates.getModule().getNameAndNoExts();
+        }
+
+        return name;
+    }
+
+    private PsiMethod findMethod(PsiClass unit, String methodName) {
+        PsiMethod[] methods = unit.getMethods();
+        if (methods == null) {
+            return null;
```
Comment (spg) [inline src/com/arcbees/plugin/idea/wizards/createpresenter/CreatePresenterAction.java:289]: 1 return statement

**Comment by spg** [comment]
Other than a few extra return statements here and there, I think CreatePresenterAction is starting to grow fast. Maybe consider splitting it smaller classes?

**Comment by branflake2267** [comment]
Yeah, I'll need to refine my returns in places to conform :). I agree about the class extraction, many of the processes could be moved to utility classes cutting down on the clutter. And larger methods can be broken down. My current strategy has been to match my first iteration with Eclipse, so the same goes for its code too, but also my main goal is down in dirty get it done and iterate better changes. Most of the time I'm flailing around with figuring out the api and haven't spent much time on boiling some down, instead aiming for get it done. Some of the extraction I believe may take testing b/c it has the possibility of changing the processes during the remodel. Anyway, I agree with you all. :)

**Comment by spg** [comment]
Yeah I work the same way too. My general approach to developing is flail around trying to make things work -> apply some polish while writing unit tests

**Comment by branflake2267** [comment]
Thats awesome. Good point.

**Comment by meriouma** [comment]
Apart from spg's comments, looks good!

**Pull request merged**